### PR TITLE
clear() -> clear_rules()

### PR DESCRIPTION
**Pull request by leina05**
PR checklist:

- [x] Added changelog entry.
- [x] All non-python libraries

### CODE REVIEW

**Comment by dhatch** [comment]
looks good so far

File: languages/ruby/lib/oso/polar/ffi/polar.rb [hunk 1]
```diff
@@ -39,6 +40,13 @@ def load(src, filename: nil)
           raise FFI::Error.get if loaded.zero?
         end
 
+        # @raise [FFI::Error] if the FFI call returns an error.
+        def clear_rules
+          cleared = Rust.clear_rules(self)
+
```
Comment (gj) [inline languages/ruby/lib/oso/polar/ffi/polar.rb:46]: Do we need to `process_messages` here?

File: languages/rust/oso/src/oso.rs [hunk 2]
```diff
@@ -64,8 +64,9 @@ impl Oso {
     }
 
     /// Clear out all files and rules that have been loaded.
-    pub fn clear(&mut self) {
-        *self = Self::new();
+    pub fn clear_rules(&mut self) -> crate::Result<()> {
+        self.inner.clear_rules()?;
```
Comment (gj) [inline languages/rust/oso/src/oso.rs:68]: `check_messages!(self.inner);` ?

File: polar-core/src/polar.rs [hunk 3]
```diff
@@ -180,6 +180,18 @@ impl Polar {
         self.load(src, None)
     }
 
+    /// Clear rules from the knowledge base
+    pub fn clear_rules(&self) -> PolarResult<()> {
+        let mut kb = self.kb.write().unwrap();
+        kb.rules = HashMap::new();
+        kb.sources = Sources::default();
+        kb.inline_queries = vec![];
```
Comment (gj) [inline polar-core/src/polar.rs:188]: Why do these three reassign and the following two call `clear()` on the existing data structure? Why not call `clear()` on all five (or reassign all five)?

File: polar-core/src/polar.rs [hunk 4]
```diff
@@ -180,6 +180,18 @@ impl Polar {
         self.load(src, None)
     }
 
+    /// Clear rules from the knowledge base
+    pub fn clear_rules(&self) -> PolarResult<()> {
```
Comment (gj) [inline polar-core/src/polar.rs:184]: Does this need to return a result? what's the error case that's getting propagated up? Unless I'm missing something, I only see panics (via the `unwrap()` calls) and success cases.

File: languages/python/oso/polar/ffi.py [hunk 5]
```diff
@@ -27,6 +27,11 @@ def load(self, string, filename=None):
         process_messages(self.next_message)
         check_result(result)
 
+    def clear_rules(self):
+        """Clear all rules from the Polar KB"""
+        result = lib.polar_clear_rules(self.ptr)
```
[the remark this answers is not in the document]
Comment (gj) [reply languages/python/oso/polar/ffi.py:32]: ETA: I'm not saying we need to -- just that it seems like if we need to in one lib it seems reasonable that we should need to in all of them

**Review by gj — CHANGES_REQUESTED** [review]
Excellent work; few questions

File: languages/python/oso/polar/ffi.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gj) [reply languages/python/oso/polar/ffi.py:32]: Reading the Rust code again -- are you sure we need to check messages here? I'm not seeing which messages would be queued by `clear_rules()` on the Rust side and in need of checking

File: polar-core/src/polar.rs (hunk 4 above)
Comment (leina05) [reply polar-core/src/polar.rs]: I changed this so it doesn't return a result. The library functions still return 1 or 0 for failure, as panics are converted to failures.

File: polar-core/src/polar.rs (hunk 3 above)
Comment (leina05) [reply polar-core/src/polar.rs]: Updated to be consistent. `Sources` still needs to be reassigned to the default.

File: languages/python/oso/tests/parity/test_api.py [hunk 6]
```diff
@@ -247,12 +247,12 @@ def test_iter_fields(polar, load_policy, query):
 
 
 @pytest.mark.xfail(EXPECT_XFAIL_PASS, reason="Test relies on internal classes.")
```
[the remark this answers is not in the document]
Comment (leina05) [reply languages/python/oso/tests/parity/test_api.py:249]: This is a super old test that uses old syntax. I wasn't using it to test this, but I can try to get it working again. The "relies on internal classes" refers to the fact that it uses classes defined in Polar.

[the remark this answers is not in the document]
Comment (leina05) [reply languages/python/oso/tests/parity/test_api.py:249]: I added a new python test to explicitly test `clear_rules()`

File: polar-wasm-api/src/polar.rs [hunk 7]
```diff
@@ -24,6 +24,12 @@ impl Polar {
             .map_err(Error::into)
     }
 
+    #[wasm_bindgen(js_class = Polar, js_name = clearRules)]
+    pub fn wasm_clear_rules(&self) -> JsResult<()> {
+        self.0.clear_rules();
+        Ok(())
+    }
```
Comment (gj) [inline polar-wasm-api/src/polar.rs:31]:
```suggestion
    pub fn wasm_clear_rules(&self) {
        self.0.clear_rules()
    }
```

Comment (leina05) [reply polar-wasm-api/src/polar.rs]: What will happen now if `clear_rules()` panics?

Comment (gj) [reply polar-wasm-api/src/polar.rs]: It'll blow up with a pretty stack trace courtesy of https://github.com/rustwasm/console_error_panic_hook

File: languages/rust/oso/src/oso.rs [hunk 8]
```diff
@@ -64,8 +64,9 @@ impl Oso {
     }
 
     /// Clear out all files and rules that have been loaded.
-    pub fn clear(&mut self) {
-        *self = Self::new();
+    pub fn clear_rules(&mut self) -> crate::Result<()> {
+        self.inner.clear_rules();
+        Ok(())
```
Comment (gj) [inline languages/rust/oso/src/oso.rs:69]:
```suggestion
    pub fn clear_rules(&mut self) {
        self.inner.clear_rules()
```

File: languages/python/oso/polar/ffi.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dhatch) [reply languages/python/oso/polar/ffi.py:32]: In general I think we should check messages after most API calls.  We could easily add logging later that would occur during clear_rules, and shouldn't have to go back in the libraries and add a call to remember to check them.

It's not particularly expensive & clear_rules would likely not occur in the critical path, so an extra call doesn't seem like an issue.

**Review by gj — APPROVED** [review]

**Review by dhatch — COMMENTED** [review]
Commented inline - not a huge deal, but I don't agree with the rationale for removing `check_messages`.  It would be best to do that check as frequently as possible so that messages are logged near the time they are generated in the core.  We may add messages that log during clear_rules in the future, and it doesn't hurt to have the extra check there.

File: languages/python/oso/polar/ffi.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (leina05) [reply languages/python/oso/polar/ffi.py:32]: Yeah that makes sense. Rebasing to add message checks back in...

**Pull request merged**